py-data-aggregation/db/aggregation_producer.cpp

```diff
@@ -0,0 +1,123 @@
+  char* iterations = std::getenv("SS_ITERATIONS");
+  int iters = iterations ? std::stoi(iterations) : 20;
+  return iters;
+}
+
+void run_aggregation_production(size_t n_bytes,
+                                size_t tensors_per_dataset)
+{
+    int rank;
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+    if (rank == 0)
+        std::cout << "Connecting clients" << std::endl;
+
+    // Connect a client for each MPI rank
+    SmartRedis::Client client(true);
+
+    // Block for all clients to be connected
+    MPI_Barrier(MPI_COMM_WORLD);
+
+    // Create tensors for the dataset
+    size_t n_values = n_bytes / sizeof(float);
+    std::vector<float> array(n_values, 0);
+    std::vector<float> result(n_values, 0);
+    for(size_t i = 0; i < n_values; i++)
+        array[i] = i;
+
+    // Get the number of iterations to perform
+    int iterations = get_iterations();
+
+    // Put the datasets into the database. The raw
+    // dataset data can be re-used between iterations
+    // because a new list is created for each iteration
+    // We re-use datasets so we don't run out of memory.
+
+    // Set the dataset name (MPI rank dependent)
+    std::string name = "aggregation_rank_" +
+                        std::to_string(rank);
+
+    // Create the dataset and add specified number of tensors
+    SmartRedis::DataSet dataset(name);
+    for (size_t j = 0; j < tensors_per_dataset; j++) {
+        std::string tensor_name = "tensor_" + std::to_string(j);
+        dataset.add_tensor(tensor_name,
+                           array.data(),
+                           {1, n_values},
+                           SRTensorTypeFloat,
+                           SRMemLayoutContiguous);
+    }
+
+    // Put the dataset into the database
+    client.put_dataset(dataset);
+
+    // A new list is created for each iteration
+    // to measure dataset aggregation throughput
+    for (int i = 0; i < iterations; i++) {
+
+        // Set the list name (not MPI rank dependent)
+        std::string list_name = "iteration_" + std::to_string(i);
+
+        // This poll invocation syncs the producer to wait for the
+        // consumer to finish the previous iteration aggregation
+        if (rank == 0 && i != 0) {
+            std::string last_list_name = "iteration_" + std::to_string(i - 1);
+            while (client.get_list_length(last_list_name) != 0) {
+                std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+            }
+        }
+
+        // Block until the consumer has deleted the pervious iteration list
+        MPI_Barrier(MPI_COMM_WORLD);
+
+        // Put and Append to the aggregation list
+        client.put_dataset(dataset);
+        client.append_to_list(list_name, dataset);
+    }
+}
+
+int main(int argc, char* argv[]) {
+
+    MPI_Init(&argc, &argv);
+
+    int rank;
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+    // Get command line arguments
+    if(argc==1)
+        throw std::runtime_error("The number tensor size in "\
```

Comment:
More standard form here would be to call a `usage()` routine that writes a usage summary to stdout and then calls `exit(-1)`.

Reply:
That's much cleaner!! I'll keep that in mind for next time!!